### PR TITLE
Generic Host PCIe Controller Support

### PR DESCRIPTION
**Pull request by superna9999**
This PR is the first step into PCIe Host Controller support on non-x86 platforms (aka Generic or platform-agnostic)

This adds :
- edtlib/gen_defines: changes to handle the ranges DT attribute
- pci-host-ecam-generic bindings
- PCIe Controller API
- Generic PCIe Controller layer implementing the current PCIe API
- Generic PCIe Controller in ECAM mode driver

The Generic PCIe Controller layer provides:
- Configuration space read/write
- single bus endpoint enumerations
- Endpoint I/O, MEM & MEM64 BARs allocation
- Endpoint I/O, MEM & MEM64 BARs get & translation for drivers

The Generic PCIe Controller in ECAM mode driver provides:
- Raw DT RANGES properties into usable PCIe regions
- Configuration space read/write into ECAM config space
- PCIe regions allocation & translation

The limitations are:
- No support for PCIe prefetchable regions
- No support for PCIe bus configuration (only bus0 is supported)
- No support for multiple controllers (no domain-id in BDF)

Support has been designed to initially support Root Complexes with Root Complex Integrated Endpoint, which was designed for Embedded Systems with internal-only PCIe Endpoints on bus 0.

Testing has been done with QEMU with endpoints only on bus0, this can be achieved with:
```
-device <device>,bus=pcie.0
```

MSI/MSI-X support depends on #37506 completion

DT RANGES tooling and macros are pushed to #38106 and this serie is rebased on it, will need to be merged before this one.

Extended PCI(e) capability offset & defines changes were moved to #38256

### CODE REVIEW

**Comment by galak** [comment]
@superna9999 do you mind pull the devicetree tooling stuff into its own PR for review and acceptance.

File: include/drivers/pcie/controller.h [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+		uintptr_t host_map_addr;
+		/* region size */
+		size_t map_length;
+	} ranges[];
+};
+
+/*
+ * Get BAR regions translation ranges count from DT
+ */
+#define DT_RANGES_COUNT(node_id) \
```
Comment (galak) [inline include/drivers/pcie/controller.h:278]: These should be in an include/devicetree/ header not here.

**Review by galak — DISMISSED** [review]
* Pull DT tooling and DT API (ie DT_RANGES_ macros) into their own PR
* Also need tests added for the DT_RANGES_ macros (however, I suggest waiting on that until agreement on the API)

**Comment by superna9999** [comment]
> * Pull DT tooling and DT API (ie DT_RANGES_ macros) into their own PR
> * Also need tests added for the DT_RANGES_ macros (however, I suggest waiting on that until agreement on the API)

Will do in 2w after my vacations

File: drivers/pcie/host/core.c [hunk 2]
```diff
@@ -0,0 +1,215 @@
+/*
```
Comment (tbursztyka) [inline drivers/pcie/host/core.c:1]: same as for the file name, controller.c would be better no?

File: drivers/pcie/host/Kconfig [hunk 3]
```diff
@@ -10,6 +10,23 @@ menuconfig PCIE
 
 if PCIE
 
+config PCIE_CORE
```
[the remark this answers is not in the document]
Comment (superna9999) [reply drivers/pcie/host/Kconfig:13]: yes, I wasn't sure about naming, will rename to that

File: drivers/pcie/host/core.c (hunk 2 above)
Comment (superna9999) [reply drivers/pcie/host/core.c:1]: Yes

File: include/drivers/pcie/msi.h [hunk 4]
```diff
@@ -111,8 +112,8 @@ extern bool pcie_msi_enable(pcie_bdf_t bdf,
  * MSI capability IDs in the PCI configuration capability list.
  */
 
-#define PCIE_MSI_CAP_ID		0x05U
-#define PCIE_MSIX_CAP_ID	0x11U
+#define PCIE_MSI_CAP_ID		PCI_CAP_ID_MSI
+#define PCIE_MSIX_CAP_ID	PCI_CAP_ID_MSIX
```
Comment (finikorg) [inline include/drivers/pcie/msi.h:116]: Can we just use `PCI_CAP_ID_MSI` instead of `PCIE_MSI_CAP_ID`.

Comment (superna9999) [reply include/drivers/pcie/msi.h:116]: Obviously yes, I wanted to minimize changes but yes those defines should be dropped

File: drivers/pcie/host/msi.c [hunk 5]
```diff
@@ -6,6 +6,7 @@
 
 #include <kernel.h>
 #include <drivers/pcie/msi.h>
+#include <drivers/pcie/cap.h>
```
Comment (tbursztyka) [inline drivers/pcie/host/msi.c:9]: Could you put these 3 last commits into their own PR? They are not specifically related to PCIe host controller support, and they could get merged faster as well.

**Review by tbursztyka — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

File: include/drivers/pcie/controller.h [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+/*
+ * Copyright (c) 2021 BayLibre, SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_DRIVERS_PCIE_CONTROLLERS_H_
+#define ZEPHYR_INCLUDE_DRIVERS_PCIE_CONTROLLERS_H_
+
+#include <zephyr/types.h>
+#include <devicetree/ranges.h>
```
Comment (mbolivar-nordic) [inline include/drivers/pcie/controller.h:16]: Just include devicetree.h. All `devicetree/foo.h` headers are private.

File: drivers/pcie/host/controller.c [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+#include <drivers/pcie/msi.h>
+#endif
+
+/* arch agnostic PCIe API implementation */
+
+uint32_t pcie_conf_read(pcie_bdf_t bdf, unsigned int reg)
+{
+	const struct device *dev;
+
+	dev = DEVICE_DT_GET(DT_ALIAS(pcie0));
```
Comment (mbolivar-nordic) [inline drivers/pcie/host/controller.c:24]: This looks like it should be a chosen node. Aliases in Zephyr are generally reserved for samples (think `led0`). Why is this an alias?

Comment (superna9999) [reply drivers/pcie/host/controller.c:24]: @mbolivar-nordic ok, I just followed what Linux does, busses can be specified as aliases, `chosen` is exclusively use by bootloaders

The DT spec goes in this side

File: include/drivers/pcie/controller.h (hunk 6 above)
Comment (superna9999) [reply include/drivers/pcie/controller.h:16]: Yes, I need to rebase on #38106

File: drivers/pcie/host/controller.c (hunk 7 above)
Comment (superna9999) [reply drivers/pcie/host/controller.c:24]: @mbolivar-nordic so according to Zephyr doc, I should a new `zephyr,pcie-controller` Zephyr-specific chosen property ?

Comment (superna9999) [reply drivers/pcie/host/controller.c]: OK done

Comment (mbolivar-nordic) [reply drivers/pcie/host/controller.c]: > @mbolivar-nordic so according to Zephyr doc, I should a new `zephyr,pcie-controller` Zephyr-specific chosen property ?

That's what we generally do when we want to pick a node for use by a subsystem, yeah. Thanks.

**Comment by superna9999** [comment]
> * Pull DT tooling and DT API (ie DT_RANGES_ macros) into their own PR
> * Also need tests added for the DT_RANGES_ macros (however, I suggest waiting on that until agreement on the API)

PR #38106 is now merged, and this was rebased on main

File: drivers/pcie/host/pcie.c [hunk 8]
```diff
@@ -142,13 +162,32 @@ bool pcie_get_mbar(pcie_bdf_t bdf,
...
 
+#ifdef CONFIG_PCIE_CONTROLLER
+	/* Translate to physical memory address from bus address */
+	if (!pcie_ctrl_region_xlate(dev, bdf, PCIE_CONF_BAR_MEM(phys_addr),
+				    PCIE_CONF_BAR_64(phys_addr),
+				    PCIE_CONF_BAR_MEM(phys_addr) ?
+					  PCIE_CONF_BAR_IO_ADDR(phys_addr)
+					: PCIE_CONF_BAR_ADDR(phys_addr),
+				    &mbar->phys_addr))
+		return false;
```
Comment (jhedberg) [inline drivers/pcie/host/pcie.c:187]: Missing { }

File: drivers/pcie/host/controller.c [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+	return bdf_cfg_mem[reg];
+}
+
+void generic_pcie_ctrl_conf_write(mm_reg_t cfg_addr, pcie_bdf_t bdf,
+				 unsigned int reg, uint32_t data)
+{
+	volatile uint32_t *bdf_cfg_mem = (volatile uint32_t *)((uintptr_t)cfg_addr + (bdf << 4));
+
+	if (!cfg_addr)
+		return;
```
Comment (jhedberg) [inline drivers/pcie/host/controller.c:62]: Missing { }

File: drivers/pcie/host/controller.c [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+	pcie_ctrl_conf_write(dev, bdf, reg, data);
+}
+
+uint32_t generic_pcie_ctrl_conf_read(mm_reg_t cfg_addr, pcie_bdf_t bdf, unsigned int reg)
+{
+	volatile uint32_t *bdf_cfg_mem = (volatile uint32_t *)((uintptr_t)cfg_addr + (bdf << 4));
+
+	if (!cfg_addr)
+		return 0xffffffff;
```
Comment (jhedberg) [inline drivers/pcie/host/controller.c:51]: Missing { }

**Review by jhedberg — CHANGES_REQUESTED** [review]
A few minor coding style issue regarding missing braces (Zephyr requires them for even single-line branches). What's more strange is that we apparently still don't have an automated check for this in CI?

File: drivers/pcie/host/pcie.c (hunk 8 above)
Comment (superna9999) [reply drivers/pcie/host/pcie.c:187]: thx

File: drivers/pcie/host/controller.c (hunk 9 above)
Comment (superna9999) [reply drivers/pcie/host/controller.c:62]: thx

File: drivers/pcie/host/controller.c (hunk 10 above)
Comment (superna9999) [reply drivers/pcie/host/controller.c:51]: thx

**Review by jhedberg — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by mbolivar-nordic — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Comment by superna9999** [comment]
> * Pull DT tooling and DT API (ie DT_RANGES_ macros) into their own PR
> * Also need tests added for the DT_RANGES_ macros (however, I suggest waiting on that until agreement on the API)

@galak is there something blocking ?

**Comment by cfriedt** [comment]
🥳
